contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,401 @@
+        }
+
+        uint256 _shares = IStakePool(pool).delegate{value: _bnbAmount}(delegator);
+        valInfo.totalShares += _shares;
+        valInfo.totalPooledBNB += _bnbAmount;
+        totalPooledBNB += _bnbAmount;
+
+        _updateEligibleValidators(validator, valInfo.totalPooledBNB, valInfo.voteAddress);
+    }
+
+    function undelegate(address validator, uint256 _sharesAmount) public onlyInitialized validatorExist(validator) {
+        require(_sharesAmount > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator storage valInfo = validators[validator];
+        address pool = valInfo.poolModule;
+
+        uint256 _bnbAmount = IStakePool(pool).undelegate(delegator, _sharesAmount);
+        valInfo.totalShares -= _sharesAmount;
+        valInfo.totalPooledBNB -= _bnbAmount;
+        totalPooledBNB -= _bnbAmount;
+
+        _updateEligibleValidators(validator, valInfo.totalPooledBNB, valInfo.voteAddress);
+    }
+
+    function redelegate(address srcValidator, address dstValidator, uint256 _sharesAmount) public onlyInitialized validatorExist(srcValidator) validatorExist(dstValidator) {
+        require(_sharesAmount > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator storage srcValInfo = validators[srcValidator];
+        Validator storage dstValInfo = validators[dstValidator];
+        address srcPool = srcValInfo.poolModule;
+        address dstPool = dstValInfo.poolModule;
+
+        uint256 _bnbAmount = IStakePool(srcPool).redelegate(delegator, _sharesAmount);
+        uint256 _newSharesAmount = IStakePool(dstPool).delegate{value: _bnbAmount}(delegator);
+
+        srcValInfo.totalShares -= _sharesAmount;
+        srcValInfo.totalPooledBNB -= _bnbAmount;
+        dstValInfo.totalShares += _newSharesAmount;
+        dstValInfo.totalPooledBNB += _bnbAmount;
+
+        _updateEligibleValidators(srcValidator, srcValInfo.totalPooledBNB, srcValInfo.voteAddress);
+        _updateEligibleValidators(dstValidator, dstValInfo.totalPooledBNB, dstValInfo.voteAddress);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address validator) external payable onlyInitialized onlyValidatorContract {
+        Validator storage valInfo = validators[validator];
+        require(!valInfo.jailed, "VALIDATOR_JAILED");
+
+        valInfo.totalPooledBNB += msg.value;
+        totalPooledBNB += msg.value;
+
+        IStakePool(valInfo.poolModule).distributeReward{value: msg.value}();
+    }
+
+    function getEligibleValidators() external view returns (IBSCValidatorSet.Validator[] memory, bytes[] memory) {
+        return (eligibleValidators, eligibleValidatorVoteAddrs);
+    }
+
+    function downtimeSlash(address validator) external onlyInitialized onlySlash {
+        Validator storage valInfo = validators[validator];
+        require(!valInfo.jailed, "VALIDATOR_JAILED");
```

Comment:
`if valInfo.jailed: return`.

Please note that every function that called by the consensus engine should success, never revert, otherwise blockchain will stop.

Slash is a function that may called by consensus engine.

Reply:
This function will be wrapped by `slashIndicator`